compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,272 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.h"
+#include "iree-dialects/Dialect/VectorExt/IR/VectorExtOps.h"
+#include "iree/compiler/Codegen/Common/GPU/VectorLayoutProvider.h"
+#include "iree/compiler/Codegen/Common/VectorLayoutAnalysis.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/Verifier.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-vector-distribution"
+
+using namespace mlir::iree_compiler::IREE::VectorExt;
+
+namespace mlir::iree_compiler {
+
+TypedValue<VectorType> getDistributed(RewriterBase &rewriter,
+                                      TypedValue<VectorType> value,
+                                      LayoutProvider *provider) {
+  // If this is a result of a "to_simd" op, use the source value of it.
+  if (auto toSIMD = value.getDefiningOp<IREE::VectorExt::ToSIMDOp>()) {
+    value = cast<TypedValue<VectorType>>(toSIMD.getInput());
+    return value;
+  }
+  // Create a "to_simt" op to convert the value to the distributed layout.
+  SmallVector<int64_t> distributedShape = provider->getDistributedShape(value);
+  VectorType distributedType =
+      VectorType::get(distributedShape, value.getType().getElementType());
+  auto toSIMT = rewriter.create<IREE::VectorExt::ToSIMTOp>(
+      value.getLoc(), distributedType, value);
+  return toSIMT.getResult();
+}
+
+void replaceOpWithDistributedValues(RewriterBase &rewriter, Operation *op,
+                                    LayoutProvider *provider,
+                                    ValueRange values) {
+  // Replace all OpResults with the given values.
+  SmallVector<Value> replacements;
+  for (OpResult opResult : op->getOpResults()) {
```

Comment:
Nit: can just do

```
for (auto [opResult, replacement] : llvm::zip_equal(op->getOpResults(), values))
```

Reply:
Upsignalling. Can you say why this wasnt addressed?

Reply:
oops, I missed this. Thanks for reminding!